index.d.ts

```diff
@@ -52,15 +43,18 @@ export interface AnyAction extends Action {
  *
  * *Do not put API calls into reducers.*
  *
- * @template S State object type.
+ * @template S The type of state consumed and produced by this reducer.
+ * @template A The type of actions the reducer can potentially respond to.
  */
-export type Reducer<S> = (state: S, action: AnyAction) => S;
+export type Reducer<S = {}, A extends Action = Action> = (state: S, action: A) => S;
 
 /**
  * Object whose values correspond to different reducer functions.
+ *
+ * @template A The type of actions the reducers can potentially respond to.
  */
-export interface ReducersMapObject {
-  [key: string]: Reducer<any>;
+export type ReducersMapObject<S = {}, A extends Action = Action> = {
+  [K in keyof S]: Reducer<S[K], A>;
```

Comment:
Already done on the `next` branch: https://github.com/reactjs/redux/blob/next/index.d.ts#L52

This PR should probably target the `next`.